### PR TITLE
Add in-memory buffer client and `get_traces` fluent API for notebook UI development

### PR DESCRIPTION
**Pull request by B-Step62**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/B-Step62/mlflow/pull/11458?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/11458/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 11458
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

This PR implements following logic as a preparation for notebook UI development:
* `InMemoryTraceClient` - a simple client that stores all traces `log_trace()` in a fixed size queue, instead of logging to the backend. This replaces the toy HTML rendering client we had for the first preview.
* `mlflow.get_traces()` - a new fluent API for retrieving N recent traces from the queue. Currently it only accepts `n` as an argument but it will support other filtering criteria like trace_id, timestamp, in the future.

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

Tested with the demo notebook.
Screenshot 2024-03-19 at 19 30 19

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/tracing/types/wrapper.py [hunk 1]
```diff
@@ -120,7 +120,8 @@ def _to_mlflow_span(self):
             end_time=self._span._end_time,
             inputs=self._inputs,
             outputs=self._outputs,
-            attributes=self._span.attributes,
+            # Convert from MappingProxyType to dict for serialization
+            attributes=dict(self._span.attributes),
```
Comment (daniellok-db) [inline mlflow/tracing/types/wrapper.py:124]: nice!

File: tests/tracing/test_types.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+def test_json_deserialization(mock_client):
+    class TestModel:
+        @mlflow.trace()
+        def predict(self, x, y):
+            z = x + y
+            z = self.add_one(z)
+            return z  # noqa: RET504
+
+        @mlflow.trace(
+            span_type=SpanType.LLM, name="add_one_with_custom_name", attributes={"delta": 1}
```
Comment (daniellok-db) [inline tests/tracing/test_types.py:16]: Could we add some un-json-serializable thing in attributes so we can test that it doesn't throw as well?

```suggestion
            span_type=SpanType.LLM, name="add_one_with_custom_name", attributes={"delta": 1, "time": datetime.now()}
```

**Review by daniellok-db — APPROVED** [review]
overall lgtm, thanks for this! it will make testing the UI a lot easier

File: tests/tracing/test_types.py (hunk 2 above)
Comment (B-Step62) [reply tests/tracing/test_types.py:16]: Ooh actually turns out attributes can only be one of these (basically serializable) in OTel....
```
Invalid type datetime for attribute 'time' value. Expected one of ['bool', 'str', 'bytes', 'int', 'float'] or a sequence of those types
```
We can support those unserializable data by some hack, but I think it's ok at this moment. We can add support anytime when it is requested from preview:)

**Pull request merged**